### PR TITLE
[feature] [null support # 1] selection only literal in broker null support

### PR DESCRIPTION
**Pull request by 61yao**
Support null literal in broker logic for selection only query. 

Scalar functions are categorized as null tolerant vs intolerant. 

intolerant scalar functions return null when any of the arg is null.

Introduce null columnDataType as passing used to pass broker response back to user.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionInvoker.java [hunk 1]
```diff
@@ -32,12 +32,17 @@
  */
 public class FunctionInvoker {
   private final Method _method;
+  // If false, the function should return null if any of its argument is null
```
Comment (zhtaoxiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionInvoker.java:35]: If `false` -> if `true`?

File: pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockBuilder.java [hunk 2]
```diff
@@ -154,6 +154,9 @@ public static RowDataBlock buildFromRows(List<Object[]> rows, DataSchema dataSch
           case OBJECT:
             setColumn(rowBuilder, byteBuffer, value);
             break;
+          case UNKNOWN:
+            setColumn(rowBuilder, byteBuffer, (Object) value);
```
Comment (zhtaoxiang) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockBuilder.java:158]: (Object) value -> (Object) null?

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionInvoker.java (hunk 1 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionInvoker.java]: Good catch. Updated the comment.

File: pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockBuilder.java (hunk 2 above)
Comment (61yao) [reply pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockBuilder.java]: The value should already be null here. but yeah, it is more clear we pas in a null. updated the code.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java [hunk 3]
```diff
@@ -230,6 +231,9 @@ private static String getStringValue(Expression thriftExpression) {
       throw new BadQueryRequestException(
           "Pinot does not support column or function on the right-hand side of the predicate");
     }
+    if (thriftExpression.getLiteral().getSetField() == Literal._Fields.NULL_VALUE) {
+      return String.valueOf(null);
```
Comment (KKcorps) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:235]: We can avoid branching here and simply return "null"

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 4]
```diff
@@ -393,6 +393,7 @@ public enum DataType {
     BOOLEAN(INT, false, true),
     TIMESTAMP(LONG, false, true),
     STRING(false, true),
+    UNKNOWN(false, true),
```
Comment (gortiz) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:396]: There are like one thousand places where FieldSpec.DataType, which is specially problematic in switch. Here the datatype I proposed in https://github.com/apache/pinot/pull/10361 may be useful, but we decided to drop it.

Have you checked the impact of this new literal in these switch? Or do you plan to do that in another PR?

**Comment by gortiz** [comment]
Can some contributor approve the workflow please? Otherwise we cannot see the result of the tests.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 4 above)
Comment (61yao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:396]: Yeah. this PR just adds the interface needed and handles the broker logic. 
Adding an extra type makes sure the code will error out when there is an error. instead of using some existing type to hide the error.

Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:396]: My point is that this PR does add `UNKNOWN` literal into `FieldSpec.DataType`, but doesn't changes any of the places where `FieldSpec.DataType` was used. This is specially important on switches. If they include a `default` case, the new literal will fall on that case, which will usually throw exceptions. In case of `switch`s without default case it can even be worse.

Comment (61yao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:396]: Oh. I''ll add the handling later since null literal isn't supported today anyway

**Review by gortiz — APPROVED** [review]
LGTM

**Review by zhtaoxiang — APPROVED** [review]

**Review by KKcorps — APPROVED** [review]

File: pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java [hunk 5]
```diff
@@ -173,7 +175,16 @@ public static Expression getLiteralExpression(byte[] value) {
     return expression;
   }
 
+  public static Expression getUnknownLiteralExpression() {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java:178]: (minor)
```suggestion
  public static Expression getNullLiteralExpression() {
```

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/CompileTimeFunctionsInvoker.java [hunk 6]
```diff
@@ -74,7 +75,12 @@ protected static Expression invokeCompileTimeFunctionExpression(@Nullable Expres
       if (functionInfo != null) {
         Object[] arguments = new Object[numOperands];
         for (int i = 0; i < numOperands; i++) {
-          arguments[i] = function.getOperands().get(i).getLiteral().getFieldValue();
+          Literal literal = function.getOperands().get(i).getLiteral();
+          if (literal.getSetField() == Literal._Fields.NULL_VALUE) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/CompileTimeFunctionsInvoker.java:79]: (minor)
```suggestion
          if (literal.isSetNullValue) {
```

File: pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java [hunk 7]
```diff
@@ -614,6 +614,9 @@ private void fillDataTableWithRandomData(DataTableBuilder dataTableBuilder,
             OBJECTS[rowId] = isNull ? null : RANDOM.nextDouble();
             dataTableBuilder.setColumn(colId, OBJECTS[rowId]);
             break;
+          case UNKNOWN:
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java:617]: (nit) Move it to the end

File: pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java [hunk 8]
```diff
@@ -743,6 +746,10 @@ private void verifyDataIsSame(DataTable newDataTable, DataSchema.ColumnDataType[
               Assert.assertEquals(ObjectSerDeUtils.deserialize(customObject), OBJECTS[rowId], ERROR_MESSAGE);
             }
             break;
+          case UNKNOWN:
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java:749]: (nit) Move it to the end

File: pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java [hunk 9]
```diff
@@ -743,6 +746,10 @@ private void verifyDataIsSame(DataTable newDataTable, DataSchema.ColumnDataType[
...
+            Object nulValue = newDataTable.getCustomObject(rowId, colId);
+            Assert.assertNull(nulValue, ERROR_MESSAGE);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java:751]: (nit)
```suggestion
            Assert.assertNull(newDataTable.getCustomObject(rowId, colId), ERROR_MESSAGE);
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java (hunk 5 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java]: done

File: pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java (hunk 7 above)
Comment (61yao) [reply pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java]: done

File: pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java (hunk 9 above)
Comment (61yao) [reply pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java]: done
